File: doc/blog/thread/trace.rb
# trace.rb
$ctr = 0
def trace (obj, key, *args)
  $ctr += 1
  idx = $ctr
  puts("#{Thread.current.object_id} call   \##{idx} #{key} #{args}")
  res = obj.send(key, *args)
  puts("#{Thread.current.object_id} return \##{idx} #{key} #{res}")
  return res
end

File: doc/blog/thread/trace.mjs
// trace.mjs
let ctr = 0;
export const trace = (fct, ...args) => {
  ctr += 1;
  const idx = ctr;
  console.log(
    `${process.pid} call   #${idx} ${fct.name} ${JSON.stringify(args)}`,
  );
  const res = fct(...args);
  console.log(
    `${process.pid} return #${idx} ${fct.name} ${JSON.stringify(res)}`,
  );
  return res;
};
export const traceAsync = async (fct, ...args) => {
  ctr += 1;
  const idx = ctr;
  console.log(
    `${process.pid} call   #${idx} ${fct.name} ${JSON.stringify(args)}`,
  );
  const res = await fct(...args);
  console.log(
    `${process.pid} return #${idx} ${fct.name} ${JSON.stringify(res)}`,
  );
  return res;
};
